fix: NetworkTransform Bitset not being properly reset if there is no delta [MTT-4275]

NoelStephensUnity opened the pull request:
**Bandwidth Consumption Fix:**
NetworkTransform's NetworkTransformState.Bitset was having bits set but never having bits "un-set" when there was no delta from the previous relative value or the delta had not crossed the threshold. This would result in any axis marked to synchronize previously being always marked to be synchronized which, in turn, would cause all marked axis to be sent anytime a delta in any of the marked axis crossed the threshold value.

On the non-authoritative side, NetworkTransform.ApplyInterpolatedNetworkStateToTransform was using the NetworkTransform.Sync<Position,RotAngle,Scale><axis> (i.e. NetworkTransform.SyncPositionX) as the determining factor as to whether it should apply all associated axis values as opposed to checking the updated NetworkTransformState.Has<Position,RotAngle,Scale><axis> to determine if it should be updated locally.

**Example of the Bitset Issue:**
- The authoritative transform detects changes to transform.position.x that exceeds the threshold
  - The transform.position.x bit flag is added to the authoritative state which is then synchronized with all non-authoritative remote instances
- Some time later, the authoritative transform detects changes to the transform.position.z and transform.localScale.y
  - The transform.position.z and transform.localScale.y bit flags are added to the authoritative state which is then synchronized with all non-authoritative remote instances
    - The state synchronizes the transform.position.x, transform.position.z, and transform.localScale.y values
- Some time later the authoritative transform only detects a change in transform.position.y that exceeds the threshold.
  - The transform.position.y bit flag is added to the authoritative state which is then synchronized with all non-authoritative remote instances
    - The state synchronizes the entire transform.position and the transform.localScale.y values.

**Authoritative Side Interpolation Fix**
This includes changes to fix the issue where the authoritative side was interpolating when it should not be.

**Client/Owner Authority Adjustments**
This includes adjustments to how we handle client-owner authority.  Instead of sending a ServerRpc which makes adjustments to a server write authority NetworkVariable, NetworkTransform now creates two versions of the NetworkVariable with one having server write permissions and the other having owner write permissions.

[<!-- Add short version of the JIRA ticket to the PR title (e.g. "feat: new shiny feature [MTT-123]") -->](https://jira.unity3d.com/browse/MTT-4275)



Pertains to #2022 and #2093, [BR-694](https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/issues/694)

## Changelog
- Fixed: WIP

## Testing and Documentation
- Includes integration tests. (WIP)

NoelStephensUnity committed:
fix

NetworkTransform was setting the value of m_Bitset but not resetting the individual bits if there was no deltas.  This would result in a gradually increasing message size over time when updating the NetworkTransform.

NoelStephensUnity committed:
update

migrating the assignment back down below the CanCommitToTransform

NoelStephensUnity committed:
fix

This resolves the issue where clients were applying every value marked to be synchronized even if the update didn't include specific axis values.

NoelStephensUnity committed:
test

This includes some fixes to prefabs that hadn't been updated to the more recent changes in NetworkTransform that limited the minimum threshold (i.e. they were set to 0 which causes issues)

NoelStephensUnity committed:
fix

Since we are now no longer sending the entire transform values, this addition prevents the non-authoritative side from making changes to the transform.

NoelStephensUnity committed:
style

removing whitespace

NoelStephensUnity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -62,118 +62,118 @@ internal struct NetworkTransformState : INetworkSerializable
 
             // 11-15: <unused>
-            private ushort m_Bitset;
+            internal ushort Bitset;
```
making internal for integration validation testing purposes.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-Bitset-not-being-properly-reset

NoelStephensUnity committed:
update

Adding some test adjustments and including rotation to interpolation measurements.

NoelStephensUnity committed:
style

removing an unused namespace.

NoelStephensUnity committed:
style

fixing issue with reference to UnityEngine's Object after I removed the system namespace.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-Bitset-not-being-properly-reset

NoelStephensUnity committed:
fix

This includes the adjustments for client authoritative mode to work without sending RPCs.

NoelStephensUnity committed:
fix

This update fixes the issue where the authoritative side was interpolating when interpolate is enabled.
Now, the authoritative side no longer interpolates.

NoelStephensUnity committed:
manual test updates

These changes make the player client authoritative within the Samples Menu->Additive Scene Loading manual test/sample.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -366,7 +366,19 @@ internal Vector3 Scale
         protected NetworkManager m_CachedNetworkManager;
 
-        private readonly NetworkVariable<NetworkTransformState> m_ReplicatedNetworkState = new NetworkVariable<NetworkTransformState>(new NetworkTransformState());
+        private readonly NetworkVariable<NetworkTransformState> m_ReplicatedNetworkStateServer = new NetworkVariable<NetworkTransformState>(new NetworkTransformState(), NetworkVariableReadPermission.Everyone, NetworkVariableWritePermission.Server);
```
We need two replicated state NetworkVariables since these have to be defined during the property declaration (to work properly).

LPLafontaineB commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -844,25 +944,61 @@ private void Awake()
+            m_CachedIsServer = IsServer;
+            m_CachedNetworkManager = NetworkManager;
+
+            CanCommitToTransform = IsServerAuthoritative() ? IsServer : IsOwner;
```
This should be added to OnGainedOwnership and OnLostOwnership to support changing ownership after spawning

NoelStephensUnity replied:
Which I need to add a test for that as well (seeing as the current tests all passed).

NoelStephensUnity committed:
update

reverting the change of m_Bitset to an internal.

NoelStephensUnity committed:
test update

updating test to account for reverting back to the private m_Bitset.

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -397,11 +414,22 @@ internal Vector3 Scale
+        }
+
+
+        protected void TryCommitTransform(Transform transformToCommit, double dirtyTime)
```
This is the same function signature as TryCommitTransformToServer, and that one just calls this one... who not just rename that one?

JesseOlmer replied:
Probably worth moving changing the `TryCommitTransformToServer` summary to something like `Prefer <see cref="TryCommitTransform">. Legacy method.`

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -397,11 +414,22 @@ internal Vector3 Scale
+        {
+            if (!CanCommitToTransform)
+            {
+                Debug.LogError($"[{name}] is trying to commit the transform without authority!");
```
NetworkLog?

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -366,7 +366,19 @@ internal Vector3 Scale
+        private readonly NetworkVariable<NetworkTransformState> m_ReplicatedNetworkStateOwner = new NetworkVariable<NetworkTransformState>(new NetworkTransformState(), NetworkVariableReadPermission.Everyone, NetworkVariableWritePermission.Owner);
+
+
+        internal NetworkVariable<NetworkTransformState> GetReplicatedNetworkState()
```
nit: Consider a getter property instead?

NoelStephensUnity replied:
Getter properties with logic are more expensive than methods?

JesseOlmer replied:
No, they're identical (and even if they weren't, this would be firmly in the micro-optimization territory). Properties ARE methods, and access to a property getter will be a callvirt, just like access to a method.

```
    {
        int result = app.MyProp;
        int res2 = app.MyMethod();
    }

    public int MyProp { get { /* ... */ } }

    public int MyMethod() { /* ... */ }
```

```
  // [17 9 - 17 33]
    IL_0015: ldloc.0      // app
    IL_0016: callvirt     instance int32 HelloWorld::get_MyProp()
    IL_001b: stloc.1      // result

    // [18 9 - 18 35]
    IL_001c: ldloc.0      // app
    IL_001d: callvirt     instance int32 HelloWorld::MyMethod()
    IL_0022: stloc.2      // res2

  .method public hidebysig specialname instance int32
    get_MyProp() cil managed

  .method public hidebysig instance int32
    MyMethod() cil managed
```

Perhaps you're thinking of the general guidance that "getters shouldn't be used to hide mutations or heavy calculations"?

NoelStephensUnity replied:
Yes, that is what I was thinking of...
Will change this to a getter.  👍

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -523,76 +546,130 @@ internal bool ApplyTransformToNetworkState(ref NetworkTransformState networkStat
+                }
+                else
+                {
+                    networkState.HasPositionX = false;
```
Is it safer (and cleaner to read) to set this unconditionally and then change it only if there is a substantial change?
```
networkState.HasPositionX = false;
if (syncposx && abovethreshold)
{
  haspositionx = true
}
```

Would safeguard against someone changing the SyncPositionX from true to false and having HasPositionX be stuck true with an old value? Not sure if that's an actual scenario that can happen - I'm not familiar enough with the code yet.

NoelStephensUnity replied:
From the non-authority side it means that by disabling SyncPositionX all updates to the x axis would cease for that instance.  Upon re-enabling SyncPositionX, the HasPositionX value would be re-synchronized (the new state would either have a bitset or not) and will be reset at that time.

For this PR I am going to say that we might just keep it the way it is.  PR #2110 has some more clean up and additional fixes that help with the readability.

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -605,7 +682,18 @@ internal bool ApplyTransformToNetworkState(ref NetworkTransformState networkStat
+        /// from having too large of a delta time between the time it was last updated and the time any new update is
+        /// sent.
+        /// </remarks>
+        /// <param name="networkState">new state to apply</param>
```
nit: these `param`s aren't adding much value - this is a private method so would recommend you just delete them

NoelStephensUnity committed:
update

Migrated the code that updates interpolators with the current value if the axis is not set into the AddInterpolatedState method as opposed to it being in the ApplyInterpolatedNetworkStateToTransform

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -889,15 +1044,35 @@ public override void OnLostOwnership()
             if (CanCommitToTransform)
             {
-                m_ReplicatedNetworkState.SetDirty(true);
+                GetReplicatedNetworkState().OnValueChanged -= OnNetworkStateChanged;
```
```suggestion
                replicatedState.OnValueChanged -= OnNetworkStateChanged;
```

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -1037,11 +1212,11 @@ public void Teleport(Vector3 newPosition, Quaternion newRotation, Vector3 newSca
             stateToSend.Rotation = newRotationEuler;
             stateToSend.Scale = newScale;
-            ApplyInterpolatedNetworkStateToTransform(stateToSend, transform);
+            //ApplyInterpolatedNetworkStateToTransform(stateToSend, transform, m_CachedNetworkManager.ServerTime.Time);
```
```suggestion
```

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-Bitset-not-being-properly-reset

NoelStephensUnity committed:
update

applying suggested changes

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-Bitset-not-being-properly-reset

NoelStephensUnity committed:
update

replacing the GetReplicatedNetworkState() method with a getter ReplicatedNetworkState property.
